modules/optimization/src/utils/ParameterMesh.C

```diff
@@ -263,82 +246,119 @@ ParameterMesh::projectToMesh(const Point & p) const
-{
-  Point a = *(elem->node_ptr(0));
-  Point b = *(elem->node_ptr(1));
-  Point c = *(elem->node_ptr(2));
-  Point d = *(elem->node_ptr(3));
-  Plane pl1(a, b, c);
-  Plane pl2(b, c, d);
-  Point trial1 = pl1.closest_point(p);
-  Point trial2 = pl2.closest_point(p);
-  if (!trial1.absolute_fuzzy_equals(trial2, TOLERANCE * TOLERANCE))
-    mooseError("Quad4 element is not coplanar");
-
-  if (elem->contains_point(trial1))
-    return trial1;
-  // Initialize min_distance to infinity
-  double min_distance = std::numeric_limits<double>::max();
-  // Initialize min_point to store the point with the smallest distance
-  Point min_point = p;
-
-  for (dof_id_type i : make_range(elem->n_edges()))
+  switch (elem.type())
   {
-    std::unique_ptr<const Elem> edge = elem->build_edge_ptr(i);
-    const Edge2 * edge_ptr = static_cast<const Edge2 *>(edge.get());
-    if (edge_ptr)
+    case EDGE2:
+    {
+      const Edge2 * edge_ptr = dynamic_cast<const Edge2 *>(&elem);
+      if (!edge_ptr)
+        mooseError("Cannot cast element to Edge2.");
+      LineSegment ls(*(edge_ptr->node_ptr(0)), *(edge_ptr->node_ptr(1)));
+      return ls.closest_point(p);
+    }
+
+    case TRI3:
     {
-      Point test_pt = closestPointonEdge(edge_ptr, p);
-      // Calculate the distance
-      Real distance = (test_pt - p).norm();
-      if (distance < min_distance)
-      // Update min_distance and min_point if the current distance is smaller
+      Point a = *(elem.node_ptr(0));
+      Point b = *(elem.node_ptr(1));
+      Point c = *(elem.node_ptr(2));
+      Plane pl(a, b, c);
+      Point trial = pl.closest_point(p);
+      if (elem.contains_point(trial))
+        return trial;
+      Real min_distance = std::numeric_limits<Real>::max();
+      Point min_point = p;
+
+      for (dof_id_type i : make_range(elem.n_edges()))
       {
-        min_distance = distance;
-        min_point = test_pt;
+        std::unique_ptr<const Elem> edge = elem.build_edge_ptr(i);
+        mooseAssert(dynamic_cast<const Edge2 *>(edge.get()),
+                    "Cannot build Edge2 element from Quad4 element.");
+        const Edge2 * edge_ptr = static_cast<const Edge2 *>(edge.get());
+        if (edge_ptr)
+        {
+          Point test_pt = closestPoint(*edge_ptr, p);
+          Real distance = (test_pt - p).norm();
+          if (distance < min_distance)
+          {
+            min_distance = distance;
+            min_point = test_pt;
+          }
+        }
       }
+      return min_point;
+    }
+    case QUAD4:
+    {
+      Point a = *(elem.node_ptr(0));
+      Point b = *(elem.node_ptr(1));
+      Point c = *(elem.node_ptr(2));
+      Point d = *(elem.node_ptr(3));
+      Plane pl1(a, b, c);
+      Plane pl2(b, c, d);
+      Point trial1 = pl1.closest_point(p);
+      Point trial2 = pl2.closest_point(p);
+      if (!trial1.absolute_fuzzy_equals(trial2, TOLERANCE * TOLERANCE))
+        mooseError("Quad4 element is not coplanar");
+
+      if (elem.contains_point(trial1))
```

Comment:
You need to check trial2!

Can you set this up to call back on tri for each plane in the quad, then it will return a bool, like found_point if it is inside the plane, false if outside but it will always return a distance.   Then you could have an early return from each tri.  

Also have tri call back on edge for the edge search.

Reply:
No? I have a check just above that confirms that trial1 and trial2 are the same point to a tolerance. If they were not then the two triangles formed from the quad would not be coplanar

Reply:
I was wrong.  Is trial1 and trial2 the distance from a plane? I thought it was the distance from the triangle face.  And if it wasn't in the triangle face it would fall back on the edge distance and make the distances different.